### DISCUSSION TITLE
PrismaClient datasource

### DISCUSSION THREAD

**Discussion by haleksandre** [prisma/prisma#9298] [Q&A]
Hi,

I am wondering if it is possible to change the datasource on the fly, in other words disconnect the client & point it to a new database/schema (Postgres) url or completely "kill" the client so that the disconnected client doesn't persist?

My use case is for integrated tests where a new db/schema is being generated for every set of tests.

In previous versions (not sure the exact version it stopped working), I used to be able to do something like this:

```prisma
// schema.prisma
datasource db {
  provider = "sqlite"
  url      = env("POSTGRES_URL")
}
```

```ts
// Somewhere in my code where I need to change the datasource i.e. tests beforeEach hook.
await prisma.$disconnect();
process.env.POSTGRES_URL = 'new url'
await prisma.$connect(); // or simply let prisma lazy connect
```

This was connecting the client to the new url as expected, but in the newer versions the datasource seems to be "hardcoded" at initialization & always point to the same datasource even after a disconnect + modify datasource + connect.

I then started exploring disconnect + reinitializing a new PrismaClient with the new datasource like so:

```ts
// Somewhere in my code where I need to change the datasource i.e. tests beforeEach hook.
await prisma.$disconnect();
process.env.POSTGRES_URL = 'new url'
prisma = new PrismaClient({
  datasources: {
    db: {
      url: process.env.POSTGRES_URL
    },
  },
}); 
```

But now I'm experiencing the "too many connection" issue. It's like even though I explicitly disconnect the previous client before initializing the new one, the old connection persists somehow. That is until I destroy/restart the node process where those are call, in this case the test server, only then does the connection truly disconnect.

If there are better ways of achieving integrated tests where the client points to a different datasource for each set of tests, I'm all ears.

Thanks for your time!

**Comment by ryands17** [comment]
@haleksandre 👋 

I setup dynamic datasources like [this](https://github.com/ryands17/nexus-auth/blob/master/prisma/prisma-test-environment.js) in my tests and it works fine without the "too many connections" issue. I also `$disconnect()` after each test suite is done.

**Reply by haleksandre** [reply]
Thanks for the quick answer with an example!

You seem to be doing something similar to my first example. I think Jest is using multiple processes (I might be wrong)? So if this is right, I'm assuming that each process creates a new client with the variable that you defined in your test environment?

I'm using Cypress so I can only spin up 1 process (test server) & then Cypress's tests send requests to that server. I have a setup / teardown hooks within the test server to modify things I need for testing, such as the prisma configs. But like mention above, the disconnect doesn't seem to entirely disconnect without killing the parent process so whether I disconnect, change the environment variable & reconnect or I disconnect then create an entirely new client the old connection persist. Could this be a bug or it's intentional?

**Reply by haleksandre** [reply]
I think i found a relevant issue [here](https://github.com/prisma/prisma/issues/9044) about this "persisting" connection.

**Reply by ryands17** [reply]
Yes that's possible.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
